awx_collection/plugins/lookup/tower_api.py

```diff
@@ -0,0 +1,202 @@
+    aliases: [query, data, filter, params]
+  expect_objects:
+    description:
+      - Error if the response does not contain either a detail view or a list view.
+    type: boolean
+    default: False
+    aliases: [expect_object]
+  expect_one:
+    description:
+      - Error if the response contains more than one object.
+    type: boolean
+    default: False
+  return_objects:
+    description:
+      - If a list view is returned, promote the list of results to the top-level of list returned.
+      - Allows using this lookup plugin to loop over objects without additional work.
+    type: boolean
+    default: True
+  return_all:
+    description:
+      - If the response is paginated, return all pages.
+    type: boolean
+    default: False
+  return_ids:
+    description:
+      - If response contains objects, promote the id key to the top-level entries in the list.
+      - Allows looking up a related object and passing it as a parameter to another module.
+      - This will convert the return to a string or list of strings depending on the number of selected items.
+    type: boolean
+    aliases: [return_id]
+    default: False
+  max_objects:
+    description:
+      - if C(return_all) is true, this is the maximum of number of objects to return from the list.
+      - If a list view returns more an max_objects an exception will be raised
+    type: integer
+    default: 1000
+
+notes:
+  - If the query is not filtered properly this can cause a performance impact.
+  - In addition, the built in threshold is 10,000 items; if the query returns more an exception will be thrown.
+"""
+
+EXAMPLES = """
+- name: Load the UI settings
+  set_fact:
+    tower_settings: "{{ lookup('awx.awx.tower_api', 'settings/ui') }}"
+
+- name: Lookup any users who are admins
+  set_fact:
+    admin_user_list_view: "{{ lookup('awx.awx.tower_api', 'users', query_params={ 'is_superuser': true }, return_objects=False) }}"
+
+- name: Lookup any users who are admins and get their objects directly
+  set_fact:
+    admin_user_object: "{{ lookup('awx.awx.tower_api', 'users', query_params={ 'is_superuser': true } ) }}"
+
+- name: Lookup the admin user and fail if there are more than one, and make sure its a list (without this, the response would be an object)
+  set_fact:
+    actual_admin_user: "{{ lookup('awx.awx.tower_api', 'users', query_params={ 'username': 'admin' }, expect_one=True, wantlist=True) }}"
+
+- name: Get just the user ID of the admin user
+  set_fact:
+    admin_user_id: "{{ lookup('awx.awx.tower_api', 'users', query_params={ 'username': 'admin' }, return_ids=True, expect_one=True) }}"
+
+- name: Create a job template with a looked up credential from a folded lookup
+  tower_job_template:
+    name: "{{ job_template_name }}"
+    credentials: >-
+        {{ lookup(
+            'awx.awx.tower_api',
+            'credentials',
+            query_params={ 'name' : credential_name },
+            return_ids=True,
+            expect_one=True,
```

Comment:
this isn't a very natural example to have `expect_one=True` and passing a list. I would like to use something that's a natural fit. The example can be whatever you want it to be.

What if, instead, call `tower_group` to create a group of all hosts that start with `foo_`? Right now it's not using the ability to return a list in a way that's not very coherent to a human.

Reply:
This is a fair comment, this was initially an attempt to lookup an assets ID which we talked about the other day. I like the group idea and will give that a try.

Reply:
So for the group example. The hosts list of tower_group is another place where it can't do ids in place of names. But I was able to come up with an example for it.